tests/battery_mod_test.cpp

```diff
@@ -0,0 +1,279 @@
+        CHECK_FALSE( med_mod.is_reloadable_with( itype_id( "battery" ) ) );
+        // Magazine is integral
+        CHECK( med_mod.magazine_integral() );
+    }
+
+    GIVEN( "tool compatible with light batteries" ) {
+        item flashlight( "flashlight" );
+        REQUIRE( flashlight.is_reloadable() );
+        REQUIRE( flashlight.is_reloadable_with( itype_id( "light_battery_cell" ) ) );
+
+        // Flashlight must be free of battery or existing mods
+        REQUIRE_FALSE( flashlight.magazine_current() );
+        REQUIRE( flashlight.toolmods().empty() );
+
+        WHEN( "medium battery mod is installed" ) {
+            med_mod.item_tags.insert( "IRREMOVABLE" );
+            flashlight.put_in( med_mod, item_pocket::pocket_type::MOD );
+
+            THEN( "tool modification is successful" ) {
+                CHECK_FALSE( flashlight.toolmods().empty() );
+
+                // FIXME: Required to fix #40948
+                CHECK_FALSE( flashlight.magazine_compatible().empty() );
+
+                // TODO: Check for +1 in item name
+                // TODO: Check other attribute changes
+            }
+
+            THEN( "medium batteries can be installed" ) {
+                // FIXME: Required to fix #40948
+                CHECK( flashlight.is_reloadable() );
+                CHECK( flashlight.is_reloadable_with( itype_id( "medium_battery_cell" ) ) );
+                CHECK( flashlight.magazine_compatible().count( itype_id( "medium_battery_cell" ) ) );
+
+                // TODO: Check item description
+            }
+
+            THEN( "light batteries no longer fit" ) {
+                CHECK_FALSE( flashlight.is_reloadable_with( itype_id( "light_battery_cell" ) ) );
+                CHECK_FALSE( flashlight.magazine_compatible().count( itype_id( "light_battery_cell" ) ) );
+            }
+        }
+    }
+}
+
+// Return true if the given item has non-empty ammo_types( conversion )
+bool has_ammo_types( const item &it, const bool conversion )
+{
+    return !it.ammo_types( conversion ).empty();
+}
+
+TEST_CASE( "ammo_types temp", "[ammo_types]" )
+{
+    SECTION( "bunch of tools" ) {
+        item flashlight( "flashlight" );
+        item hotplate( "hotplate" );
+        item vac_sealer( "vac_sealer" );
+        item forge( "forge" );
+        item cordless_drill( "cordless_drill" );
+
+        CHECK_FALSE( flashlight.ammo_types( false ).empty() );
+        CHECK_FALSE( hotplate.ammo_types( false ).empty() );
+        CHECK_FALSE( vac_sealer.ammo_types( false ).empty() );
+        CHECK_FALSE( forge.ammo_types( false ).empty() );
+        CHECK_FALSE( cordless_drill.ammo_types( false ).empty() );
+
+        CHECK_FALSE( flashlight.ammo_types( true ).empty() );
+        CHECK_FALSE( hotplate.ammo_types( true ).empty() );
+        CHECK_FALSE( vac_sealer.ammo_types( true ).empty() );
+        CHECK_FALSE( forge.ammo_types( true ).empty() );
+        CHECK_FALSE( cordless_drill.ammo_types( true ).empty() );
```

Comment:
`ammo_types` should return an empty set for all of these. We need to separate the concepts of what ammotypes an item can contain from what ammotypes it can use. `ammo_types` should handle the first so it should lose the `conversion` checks as those aren't applicable to pocket `ammo_restriction`s.

Reply:
Shall I remove the `conversion` argument from `item::ammo_types` entirely? I think that would leave only:

```
std::set<ammotype> item::ammo_types() const
{
    return contents.ammo_types();
}
```

Reply:
Yeah that should all go. The whole function can go if it doesn't do anything different to `item_contents::ammo_types`.

Reply:
Great - I love an opportunity to delete code.

Reply:
Now I have doubts. If `item::ammo_types` no longer accepts a `conversion` argument, then it looks like that argument would need removal from `ammo_default`, `common_ammo_defaul`, `magazine_default` and `magazine_compatible` as well. Is that what you are intending?

Reply:
All of that will need refactoring later. It could stay for now if other things are using it but if behavior in those functions depends on `ammo_types` accounting for conversions I'd expect those behaviors to be broken now.